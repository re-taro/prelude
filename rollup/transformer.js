// @ts-check

import ts from "typescript";

/**
 *
 * @param {import('typescript').Program} _program
 * @returns {import('typescript').TransformerFactory<import('typescript').SourceFile>}
 */
export function importMetaVitestTransformer(_program) {
	/**
	 * @param {import('typescript').TransformationContext} context
	 * @returns {import('typescript').Transformer<import('typescript').SourceFile>}
	 */
	return (context) => {
		/**
		 * @param {import('typescript').Node} node
		 * @returns {import('typescript').VisitResult<import('typescript').Node>}
		 */
		const visitor = (node) => {
			if (
				ts.isIfStatement(node)
				&& ts.isPropertyAccessExpression(node.expression)
				&& ts.isMetaProperty(node.expression.expression)
				&& node.expression.expression.keywordToken === ts.SyntaxKind.ImportKeyword
				&& node.expression.name.text === "vitest"
			) {
				return ts.factory.createNotEmittedStatement(node);
			}

			return ts.visitEachChild(node, visitor, context);
		};

		/**
		 * @param {import('typescript').SourceFile} sourceFile
		 * @returns {import('typescript').SourceFile}
		 */
		return (sourceFile) => {
			const updatedSourceFile = ts.visitNode(sourceFile, visitor);
			if (!ts.isSourceFile(updatedSourceFile)) {
				throw new Error("Transformation did not return a SourceFile.");
			}
			return updatedSourceFile;
		};
	};
}
