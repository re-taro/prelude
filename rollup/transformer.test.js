import { describe, expect, it } from "vitest";
import ts from "typescript";
import { importMetaVitestTransformer } from "./transformer.js";

describe("importMetaVitestTransformer", () => {
	it("should return a transformer factory", () => {
		const sourceCode = `
if (import.meta.vitest) {
  console.log("This is test code.");
}
    `.trim();

		const result = ts.transpileModule(sourceCode, {
			compilerOptions: { target: ts.ScriptTarget.ESNext },
			transformers: { before: [importMetaVitestTransformer()] },
		});

		expect(result.outputText).toMatchSnapshot();
	});
});
