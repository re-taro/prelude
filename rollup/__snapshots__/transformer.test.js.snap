// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`importMetaVitestTransformer > should return a transformer factory 1`] = `
"export {};
"
`;
